### PR TITLE
Feature/sg 849 add replace in channels

### PR DESCRIPTION
**Pull request by Louis-Dupont**
Adding `replace_input_channels ` and `get_input_channels` to `SgModel` (and other required stems/backbones)

### CODE REVIEW

File: src/super_gradients/conversion/conversion_utils.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
     (torch.int16, np.int16),
     (torch.int8, np.int8),
     (torch.uint8, np.uint8),
-    (torch.bool, np.bool),
+    # (torch.bool, np.bool),
```
Comment (BloodAxe) [inline src/super_gradients/conversion/conversion_utils.py:15]: ?

File: src/super_gradients/module_interfaces/module_interfaces.py [hunk 2]
```diff
@@ -69,3 +70,29 @@ def replace_num_classes(self, num_classes: int, compute_new_weights_fn: Callable
         :return: None
         """
         raise NotImplementedError(f"replace_num_classes is not implemented in the derived class {self.__class__.__name__}")
+
+
+class SupportsReplaceInChannels(ABC):
```
Comment (BloodAxe) [inline src/super_gradients/module_interfaces/module_interfaces.py:75]: Maybe use full name: In -> Input

File: src/super_gradients/conversion/conversion_utils.py (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/conversion/conversion_utils.py:15]: Yeah, I'll take it out! I just had an error locally and didn't have internet to update my versions, so I commented out and accidentally added it to the PR

File: src/super_gradients/module_interfaces/module_interfaces.py (hunk 2 above)
Comment (Louis-Dupont) [reply src/super_gradients/module_interfaces/module_interfaces.py]: [3be5c64](https://github.com/Deci-AI/super-gradients/pull/1557/commits/3be5c64ac670dc5a27e8383193acaeb8272e3aa8)

Comment (Louis-Dupont) [reply src/super_gradients/module_interfaces/module_interfaces.py]: I hesitate `SupportsReplaceInputChannels` -> `SupportsInputChannels` or something like that, because I added `get_input_channels` to this abstract mixin, so it's not only "replace".
But at the same time, `SupportsInputChannels` is a very weird name. So maybe keeping `SupportsReplaceInputChannels` is better, `get_input_channels` being implicit in the name, but still explicit in the abstract class definition.

File: src/super_gradients/conversion/conversion_utils.py (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/conversion/conversion_utils.py]: Removed 
8d654eb

File: src/super_gradients/modules/backbone_replacement_utils.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+from typing import Union, Optional, Callable
+
+import torch
+from torch import nn
+
+__all__ = ["replace_input_channels_with_random_weights"]
+
+
+def compute_new_weights(
+    module: Union[nn.Conv2d, nn.Linear, nn.Module],
```
Comment (shaydeci) [inline src/super_gradients/modules/backbone_replacement_utils.py:10]: Dont think nn.Linear belongs here.
Actually, nn.Module is sufficient since fn is passed.

File: src/super_gradients/modules/backbone_replacement_utils.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    in_channels: int,
+    fn: Optional[Callable[[nn.Module, int], nn.Module]] = None,
+) -> nn.Module:
+    fn = fn or replace_input_channels_with_random_weights
+    return fn(module=module, in_channels=in_channels)
+
+
+def replace_input_channels_with_random_weights(module: Union[nn.Conv2d, nn.Linear, nn.Module], in_channels: int) -> nn.Module:
```
Comment (shaydeci) [inline src/super_gradients/modules/backbone_replacement_utils.py:18]: So the typing supports nn.Linear but throws an error if it isnt conv?
It makes more sense to simply support a dedicated conv2d replacement by changing the name of this method, and dropping the ifs and exceptions.

File: tests/unit_tests/test_models_factory.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+class DynamicModelTests(unittest.TestCase):
+    def test_models(self):
+        # TODO: replace `MODELS` with `ARCHITECTURES.keys()` once all models can be instantiated with
+        # TODO  models.get() without explicit arch_params without any explicit arch_params
+        for model_name in MODELS:
+            with self.subTest(model_name=model_name):
+                model = models.get(model_name, num_classes=20)
+
+                model.replace_input_channels(3)
```
Comment (shaydeci) [inline tests/unit_tests/test_models_factory.py:110]: I think this is not sufficient.
IMO its very important to check that we can pass through the models inputs with the new channels.
I think this needs to be tested both for pretrained models in particular.

**Review by shaydeci — CHANGES_REQUESTED** [review]

File: src/super_gradients/training/models/sg_module.py [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
+from typing import Union, Optional, Callable
 
 from torch import nn
 
 from super_gradients.training.utils.utils import HpmStruct
+from super_gradients.module_interfaces import SupportsReplaceInputChannels
 
 
-class SgModule(nn.Module):
+class SgModule(nn.Module, SupportsReplaceInputChannels):
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/models/sg_module.py:9]: Originally I did not have this `SupportsReplaceInputChannels `, but there are some models that are not `SgModule` but that still need to support `replace_input_channels`, like backbones and stems.
Having an external `SupportsReplaceInputChannels` solves the problem, because we can make `SgModule` inherit from it, but also any other backbone/stem that is not `SgModule`

File: src/super_gradients/modules/backbone_replacement_utils.py (hunk 4 above)
Comment (Louis-Dupont) [reply src/super_gradients/modules/backbone_replacement_utils.py:18]: > So the typing supports nn.Linear but throws an error if it isnt conv?
That was a mistake, `nn.Linear` shouldnt be in typing

> It makes more sense to simply support a dedicated conv2d replacement by changing the name of this method, and dropping the ifs and exceptions.
Actually that's a good point, i'll change the name to be specifically on conv2d

File: src/super_gradients/modules/backbone_replacement_utils.py (hunk 3 above)
Comment (Louis-Dupont) [reply src/super_gradients/modules/backbone_replacement_utils.py:10]: (see next comment)

File: src/super_gradients/module_interfaces/module_interfaces.py (hunk 2 above)
Comment (BloodAxe) [reply src/super_gradients/module_interfaces/module_interfaces.py]: `SupportsReplaceInputChannels` sounds absolutely fine to my taste (Much better than SupportsInputChannels).
And having get_input_channels is part of mixin is also fine since it's part of the contract one must implement to support this mixin. So the design is great here, IMO.

File: src/super_gradients/modules/backbone_replacement_utils.py (hunk 4 above)
Comment (Louis-Dupont) [reply src/super_gradients/modules/backbone_replacement_utils.py]: Changed here [622c61e](https://github.com/Deci-AI/super-gradients/pull/1557/commits/622c61ebaabf752d3f0081a1d738a5924d3f67f1)

File: src/super_gradients/module_interfaces/module_interfaces.py [hunk 7]
```diff
@@ -69,3 +70,34 @@ def replace_num_classes(self, num_classes: int, compute_new_weights_fn: Callable
         :return: None
         """
         raise NotImplementedError(f"replace_num_classes is not implemented in the derived class {self.__class__.__name__}")
+
+
+class SupportsReplaceInputChannels(ABC):
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/module_interfaces/module_interfaces.py:75]: See next comment

File: src/super_gradients/modules/backbone_replacement_utils.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+from typing import Union, Optional, Callable
+
+import torch
+from torch import nn
+
+__all__ = ["replace_input_channels_with_random_weights"]
+
+
+def compute_new_weights(
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/modules/backbone_replacement_utils.py]: I renamed the function + added docs [622c61e](https://github.com/Deci-AI/super-gradients/pull/1557/commits/622c61ebaabf752d3f0081a1d738a5924d3f67f1)

File: src/super_gradients/modules/backbone_replacement_utils.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            kernel_size=module.kernel_size,
+            stride=module.stride,
+            padding=module.padding,
+            dilation=module.dilation,
+            groups=module.groups,  # Be cautious: if in_channels % groups != 0, it will raise an error.
+            bias=module.bias is not None,
+            device=module.weight.device,
+            dtype=module.weight.dtype,
+        )
+        torch.nn.init.normal_(new_module.weight, mean=module.weight.mean().item(), std=module.weight.std().item())
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/modules/backbone_replacement_utils.py]: I like that ^^

**Comment by shaydeci** [comment]
> Adding `replace_input_channels ` and `get_input_channels` to `SgModel` (and other required stems/backbones)
> 
> ### Question
> How to handle models that were implemented with `self.in_channels` attribute, since we now have `get_input_channels()` method that dynamically checks the value without relying on an attribute?
> 
> **a.** Should we remove any `self.in_channels` attribute?
> 
> pro: cleaner IMO con: risk of not being backward compatible if models are built on top of these models.
> 
> **b.** Or should we update the value of `self.in_channels` when calling `replace_input_channels()` ? pro: 100% backward compatible con: risk of bug if we miss one place where `self.in_channels` should be updated con: redundant and maybe a bit more confusing

Most important thing this PR is missin, is a mechanism to replace the stem out of the box when using pretrained weights.
One should be able to use COCO pretrained checkpoints, replacing the stem behind the scenes.

File: src/super_gradients/modules/backbone_replacement_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/modules/backbone_replacement_utils.py]: Ok, I did that but is a slightly different manner, I think it's clearer [3d114af](https://github.com/Deci-AI/super-gradients/pull/1557/commits/3d114af7f9a53c1f61dacfd75c14a8f356632c50)

File: tests/unit_tests/test_models_factory.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    "dekr_w32_no_dc",
+    "yolo_nas_pose_n",
+    "yolo_nas_pose_s",
+    "yolo_nas_pose_m",
+    "yolo_nas_pose_l",
+]
+
+
+class DynamicModelTests(unittest.TestCase):
+    def test_models(self):
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply tests/unit_tests/test_models_factory.py:118]: (same as other comment)

> I added the forward call [4afccd0](https://github.com/Deci-AI/super-gradients/pull/1557/commits/4afccd014d48e06250c06c8a19907002024ecb6c)
> I also added it to pretrained_models tests [d54f268](https://github.com/Deci-AI/super-gradients/pull/1557/commits/d54f26837acfb021a17c08005c4704c832de8198). I added the pretrained one to the integration tests because it's too slow to run in the unittests

**Comment by Louis-Dupont** [comment]
> `[287b28f](/Deci-AI/super-gradients/pull/1557/commits/287b28fbea3d22f8047dbfb07a7b40a0bac5c785) `
> [ ](/Louis-Dupont)

Added here [287b28f](https://github.com/Deci-AI/super-gradients/pull/1557/commits/287b28fbea3d22f8047dbfb07a7b40a0bac5c785)

**Comment by Louis-Dupont** [comment]
Added update of `self.in_channels` when calling `replace_input_channels`: [ab760e8](https://github.com/Deci-AI/super-gradients/pull/1557/commits/ab760e8db17131902ddf8b915d8af14aaa356ada)

File: src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py [hunk 11]
```diff
@@ -157,7 +159,7 @@ def __init__(self, in_channels: int, out_channels: int):
         :param in_channels: Number of input channels
         :param out_channels: Number of output channels
         """
-        super().__init__(in_channels)
+        super().__init__(in_channels)  # FIXME: handle self.in_channels (?)
```
Comment (BloodAxe) [inline src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py:162]: Leftover?

File: src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py [hunk 12]
```diff
@@ -168,6 +170,13 @@ def out_channels(self):
     def forward(self, x: Tensor) -> Tensor:
         return self.conv(x)
 
+    def replace_input_channels(self, in_channels: int, compute_new_weights_fn: Optional[Callable[[nn.Module, int], nn.Module]] = None):
+        # TODO: check if it makes sens to repalce the whole `QARepVGGBlock` - I think yes
```
Comment (BloodAxe) [inline src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py:174]: Still on TODO?

File: src/super_gradients/training/models/model_factory.py [hunk 13]
```diff
@@ -189,32 +189,35 @@ def get_model_name(model: torch.nn.Module) -> Optional[str]:
...
     :param checkpoint_num_classes:  num_classes of checkpoint_path/ pretrained_weights, when checkpoint_path is not None.
-     Used when num_classes != checkpoint_num_class. In this case, the module will be initialized with checkpoint_num_class, then weights will be loaded. Finaly
-        replace_head(new_num_classes=num_classes) is called (useful when wanting to perform transfer learning, from a checkpoint outside of
-         then ones offered in SG model zoo).
-
+                                        Used when num_classes != checkpoint_num_class. In this case, the module will be initialized with checkpoint_num_class,
+                                        then weights will be loaded.
+                                        Finaly replace_head(new_num_classes=num_classes) is called (useful when wanting to perform transfer learning,
+                                        from a checkpoint outside of then ones offered in SG model zoo).
+    :param num_input_channels:      Number of input channels.
```
Comment (BloodAxe) [inline src/super_gradients/training/models/model_factory.py:219]: I beleive this parameter should be introduced in checkpoint params and forwarded to models.get every time we call models.get (Multiple places in Trainer, train_from_recipe etc)

File: tests/unit_tests/test_models_factory.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply tests/unit_tests/test_models_factory.py:118]: Nice

File: src/super_gradients/training/models/model_factory.py (hunk 13 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/models/model_factory.py:219]: Added here [9d86c3e](https://github.com/Deci-AI/super-gradients/pull/1557/commits/9d86c3edded9ac584764f8e8250b93f839de68f0) [0a38118](https://github.com/Deci-AI/super-gradients/pull/1557/commits/0a381185d2335b97c0bd0f800032840d4467ca61) [a5e3c86](https://github.com/Deci-AI/super-gradients/pull/1557/commits/a5e3c86b455d7227c1fe38b861afc0a6819c1929)

File: src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py (hunk 11 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py]: removed

File: src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py (hunk 12 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/models/detection_models/yolo_nas/yolo_stages.py]: removed

File: src/super_gradients/modules/conv_bn_act_block.py [hunk 14]
```diff
@@ -67,6 +68,14 @@ def __init__(
...
+    def replace_input_channels(self, in_channels: int, compute_new_weights_fn: Optional[Callable[[nn.Module, int], nn.Module]] = None):
+        from super_gradients.modules.weight_replacement_utils import replace_conv2d_input_channels
+
+        self.seq[0] = replace_conv2d_input_channels(conv=self.seq[0], in_channels=in_channels, fn=compute_new_weights_fn)
+
+    def get_input_channels(self) -> int:
+        return self.seq[0].in_channels
+
 
 class Conv(nn.Module):
```
Comment (BloodAxe) [inline src/super_gradients/modules/conv_bn_act_block.py:80]: Missing `, SupportsReplaceInputChannels` mixin here?

Comment (Louis-Dupont) [reply src/super_gradients/modules/conv_bn_act_block.py:80]: I'm confused, it looks like it does inherit from it: `class ConvBNAct(nn.Module, SupportsReplaceInputChannels):`
Or are you talking about something else?

Comment (shaydeci) [reply src/super_gradients/modules/conv_bn_act_block.py:80]: He's talking of class Conv(nn.Module):

Comment (Louis-Dupont) [reply src/super_gradients/modules/conv_bn_act_block.py:80]: oh

Comment (Louis-Dupont) [reply src/super_gradients/modules/conv_bn_act_block.py]: [503129f](https://github.com/Deci-AI/super-gradients/pull/1557/commits/503129f8cdee297a3c555316f481e7a4502ca62c)

**Review by shaydeci — APPROVED** [review]
LGTM
